File: app/src/main/java/com/tanyayuferova/franklin/utils/FragmentFactory.java
package com.tanyayuferova.franklin.utils;

import android.support.v4.app.Fragment;

import com.tanyayuferova.franklin.ui.OnBoardingFragment;
import com.tanyayuferova.franklin.ui.ResultsFragment;
import com.tanyayuferova.franklin.ui.SettingsFragment;
import com.tanyayuferova.franklin.ui.VirtuesFragment;

/**
 * Created by Tanya Yuferova on 3/18/2018.
 */

public class FragmentFactory {

    public static Fragment createFragment(String screenKey, Object data) {
        switch (screenKey) {
            case VirtuesFragment.SCREEN_KEY:
                return VirtuesFragment.newInstance(data);
            case SettingsFragment.SCREEN_KEY:
                return SettingsFragment.newInstance(data);
            case OnBoardingFragment.SCREEN_KEY:
                return OnBoardingFragment.newInstance(data);
            case ResultsFragment.SCREEN_KEY:
                return ResultsFragment.newInstance(data);
            default:
                throw new RuntimeException("Unknown screen key");
        }
    }
}

File: app/src/main/java/com/tanyayuferova/franklin/navigation/FragmentFactory.kt
package com.tanyayuferova.franklin.navigation

import android.support.v4.app.Fragment

import com.tanyayuferova.franklin.ui.onboarding.OnBoardingFragment
import com.tanyayuferova.franklin.ui.results.ResultsFragment
import com.tanyayuferova.franklin.ui.settings.SettingsFragment
import com.tanyayuferova.franklin.ui.virtues.VirtuesFragment

/**
 * Created by Tanya Yuferova on 3/18/2018.
 */

class FragmentFactory {

    fun createFragment(screenKey: String, data: Any?): Fragment = when (screenKey) {
        VirtuesFragment.SCREEN_KEY -> VirtuesFragment.newInstance(data)
        SettingsFragment.SCREEN_KEY -> SettingsFragment.newInstance(data)
        OnBoardingFragment.SCREEN_KEY -> OnBoardingFragment.newInstance(data)
        ResultsFragment.SCREEN_KEY ->  ResultsFragment.newInstance(data)
        else -> throw RuntimeException("Unknown screen key")
    }
}